packages/flutter_tools/lib/src/android/gradle_errors.dart

```diff
@@ -669,3 +670,44 @@ final GradleHandledError sslExceptionHandler = GradleHandledError(
   },
   eventLabel: 'ssl-exception-tag-mismatch',
 );
+
+final RegExp _unsupportedClassFileMajorVersionPattern = RegExp(r'Unsupported class file major version ([0-9]+)');
+
+@visibleForTesting
+final GradleHandledError incompatibleJavaAndGradleVersionsHandler = GradleHandledError(
+  test: (String line) {
+    return _unsupportedClassFileMajorVersionPattern.hasMatch(line);
+  },
+  handler: ({
+    required String line,
+    required FlutterProject project,
+    required bool usesAndroidX,
+    required bool multidexEnabled,
+  }) async {
+    final File gradlePropertiesFile = project.directory
+        .childDirectory('android')
+        .childDirectory('gradle')
+        .childDirectory('wrapper')
+        .childFile('gradle-wrapper.properties');
+    final File gradleFile = project.directory
+        .childDirectory('android')
+        .childFile('build.gradle');
+    globals.printBox(
+      "${globals.logger.terminal.warningMark} Your project's Gradle and AGP version "
+      'is incompatible with the Java version that Flutter uses.\n\n'
+      'To fix this issue, first, check the Java version used by Flutter by '
+      'running `flutter doctor -v`.\n\n'
```

Comment:
Nit can we use the long flag names in documentation.  Shorthand is good for people typing but long flags are better for understanding. For example it would be reasonable for someone to think `-v` was version not verbose. 

```suggestion
      'running `flutter doctor --verbose`.\n\n'
```